crates/matrix-sdk-common/src/deserialized_responses.rs

```diff
@@ -432,103 +396,195 @@ impl From<DecryptedRoomEvent> for SyncTimelineEvent {
 }
 
 impl TimelineEvent {
     /// Create a new `TimelineEvent` from the given raw event.
     ///
-    /// This is a convenience constructor for when you don't need to set
-    /// `encryption_info` or `push_action`, for example inside a test.
+    /// This is a convenience constructor for a plaintext event when you don't
+    /// need to set `push_action`, for example inside a test.
     pub fn new(event: Raw<AnyTimelineEvent>) -> Self {
         Self {
-            inner_event: event,
-            inner_encryption_info: None,
+            // This conversion is unproblematic since a `SyncTimelineEvent` is just a
+            // `TimelineEvent` without the `room_id`. By converting the raw value in
+            // this way, we simply cause the `room_id` field in the json to be
+            // ignored by a subsequent deserialization.
+            inner: TimelineEventInner::PlainText { event: event.cast() },
             push_actions: None,
-            unsigned_encryption_info: None,
         }
     }
 
     /// Returns a reference to the (potentially decrypted) Matrix event inside
     /// this `TimelineEvent`.
     pub fn event(&self) -> &Raw<AnySyncTimelineEvent> {
-        // TODO: make `inner_event` an AnySyncTimelineEvent instead.
-        self.inner_event.cast_ref()
+        self.inner.event()
     }
 
     /// If the event was a decrypted event that was successfully decrypted, get
     /// its encryption info. Otherwise, `None`.
     pub fn encryption_info(&self) -> Option<&EncryptionInfo> {
-        self.inner_encryption_info.as_ref()
+        self.inner.encryption_info()
     }
 
     /// Takes ownership of this `TimelineEvent`, returning the (potentially
     /// decrypted) Matrix event within.
     pub fn into_event(self) -> Raw<AnySyncTimelineEvent> {
-        // TODO: make `inner_event` an AnySyncTimelineEvent instead.
-        self.inner_event.cast()
+        self.inner.into_event()
     }
 }
 
 impl From<DecryptedRoomEvent> for TimelineEvent {
     fn from(decrypted: DecryptedRoomEvent) -> Self {
-        Self {
-            // Casting from the more specific `AnyMessageLikeEvent` (i.e. an event without a
-            // `state_key`) to a more generic `AnyTimelineEvent` (i.e. one that may contain
-            // a `state_key`) is safe.
-            inner_event: decrypted.event.cast(),
-            inner_encryption_info: Some(decrypted.encryption_info),
-            push_actions: None,
-            unsigned_encryption_info: decrypted.unsigned_encryption_info,
-        }
+        Self { inner: TimelineEventInner::Decrypted(decrypted), push_actions: None }
     }
 }
 
 #[cfg(not(tarpaulin_include))]
 impl fmt::Debug for TimelineEvent {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        let TimelineEvent {
-            inner_event,
-            inner_encryption_info,
-            push_actions,
-            unsigned_encryption_info,
-        } = self;
+        let TimelineEvent { inner, push_actions } = self;
```

Comment:
ditto, could we just derive `Debug` on `TimelineEvent` now?

Reply:
fixed in 6e40211af